logging/zerolog/logger.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2022 CloudWeGo Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zerolog
+
+import (
+	"context"
+	"errors"
+
+	"github.com/cloudwego/hertz/pkg/common/hlog"
+	hertzzerolog "github.com/hertz-contrib/logger/zerolog"
+	"github.com/rs/zerolog"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type Logger struct {
+	hertzzerolog.Logger
+	config *config
+}
+
+// Ref to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/README.md#json-formats
+const (
+	TraceIDKey    = "trace_id"
+	SpanIDKey     = "span_id"
+	TraceFlagsKey = "trace_flags"
+)
+
+type ExtraKey string
+
+func NewLogger(opts ...Option) *Logger {
+	config := defaultConfig()
+
+	// apply options
+	for _, opt := range opts {
+		opt.apply(config)
+	}
+	logger := *config.logger
+	newLogger := hertzzerolog.From(logger.Unwrap().Hook(zerolog.HookFunc(func(e *zerolog.Event, level zerolog.Level, message string) {
+		ctx := e.GetCtx()
+		e.Any(TraceIDKey, ctx.Value(ExtraKey(TraceIDKey)))
+		e.Any(SpanIDKey, ctx.Value(ExtraKey(SpanIDKey)))
+		e.Any(TraceFlagsKey, ctx.Value(ExtraKey(TraceFlagsKey)))
+	})))
+
+	return &Logger{
+		Logger: *newLogger,
+		config: config,
+	}
+}
+
+func (l *Logger) CtxLogf(level hlog.Level, ctx context.Context, format string, kvs ...any) {
+	var zlevel zerolog.Level
+	span := trace.SpanFromContext(ctx)
+
+	ctx = context.WithValue(ctx, ExtraKey(TraceIDKey), span.SpanContext().TraceID())
+	ctx = context.WithValue(ctx, ExtraKey(SpanIDKey), span.SpanContext().SpanID())
+	ctx = context.WithValue(ctx, ExtraKey(TraceFlagsKey), span.SpanContext().TraceFlags())
+
+	switch level {
+	case hlog.LevelDebug, hlog.LevelTrace:
+		zlevel = zerolog.DebugLevel
+		l.Logger.CtxDebugf(ctx, format, kvs...)
+	case hlog.LevelInfo:
+		zlevel = zerolog.InfoLevel
+		l.Logger.CtxInfof(ctx, format, kvs...)
+	case hlog.LevelNotice, hlog.LevelWarn:
+		zlevel = zerolog.WarnLevel
+		l.Logger.CtxWarnf(ctx, format, kvs...)
+	case hlog.LevelError:
+		zlevel = zerolog.ErrorLevel
+		l.Logger.CtxErrorf(ctx, format, kvs...)
+	case hlog.LevelFatal:
+		zlevel = zerolog.FatalLevel
+		l.Logger.CtxFatalf(ctx, format, kvs...)
+	default:
+		zlevel = zerolog.WarnLevel
+		l.Logger.CtxWarnf(ctx, format, kvs...)
+	}
+
+	if !span.IsRecording() {
+		l.Logger.Logf(level, format, kvs...)
+		return
+	}
+
+	// set span status
+	if zlevel >= l.config.traceConfig.errorSpanLevel {
```

Comment:
The log level of zerolog needs more evaluation.

Reply:
I think this check is OK. By default, we use ErrorLevel as errorSpanLevel.